### PR TITLE
Aux labels plus notes on Python interface

### PR DESCRIPTION
**Pull request by danpovey**

### CODE REVIEW

File: notes/python.txt [hunk 1]
```diff
@@ -0,0 +1,87 @@
+
+  These are some notes regarding how we'll interact with Python and PyTorch.
+
+
+
+    # For composition we can rely on PyTorch's inbuilt autograd
+    def Compose(a: FsaVec, a_weights: Tensor,
+                a: FsaVec, b_weights: Tensor):
```
Comment (qindazhu) [inline notes/python.txt:8]: I guess it is `b:FsaVec, b_weights:Tensor`

File: notes/python.txt [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        else:
+           return b, b_weights, fsa.MapAuxSymbols(a_aux_symbols, indexes1, indexes2)
+
+
+
+
+    # Composing with transducers.  Assumes that A is an acceptor but B may
+    # have auxiliary symbols.
+    def TransducerCompose(a: FsaVec, a_weights: Tensor,
+                          a: FsaVec, b_weights: Tensor,
```
Comment (qindazhu) [inline notes/python.txt:51]: again, I guess it is `b:FsaVec`

File: notes/python.txt [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        if b_aux_symbols is None:
+          return c, c_weights
+        else:
+          return c, c_weights, fsa.MapAuxSymbols(b_aux_symbols, indexes_b)
+
+
+    class TotalWeight(Function):
+       """
+       Returns the total weight of FSAs (i.e. the log-sum-exp across
+       paths) as a Tensor with shapee (num_fsas,)
```
Comment (qindazhu) [inline notes/python.txt:65]: typo `shape`

File: k2/csrc/aux_labels.h [hunk 4]
```diff
@@ -50,28 +50,28 @@ struct AuxLabels {
 /*
   Maps auxiliary labels after an FSA operation where each arc in the output
   FSA corresponds to exactly one arc in the input FSA.
-     @param [in] labels_in   Labels on the arcs of the input FSA
+     @param [in] labels_in   int32_ts on the arcs of the input FSA
```
Comment (qindazhu) [inline k2/csrc/aux_labels.h:53]: the `int32_ts` means the plural of `int32_t`? (just a question as I really feel hard to get the meaning of `Auxint32_ts` from its name)

File: k2/csrc/aux_labels.h [hunk 5]
```diff
@@ -34,7 +34,7 @@ namespace k2 {
   This allows you to store auxiliary labels (e.g. olabels or ilabels)
   on each arc of an Fsa.
  */
-struct AuxLabels {
+struct Auxint32_ts {
```
Comment (csukuangfj) [inline k2/csrc/aux_labels.h:37]: This does not follow ClassNamingStyle.

File: k2/csrc/determinize.cc [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+                      // the arc at `arc_index`.
+
+  // `prev` and `next` form the doubly linked list of DetStateElement
+  DetStateElement *prev = nullptr;
+  DetStateElement *next = nullptr;
+
+  // This comparator function compares the weights, but is careful in case of
+  // ties to ensure deterministic behavior.
+  bool operator < (const DetStateElement &other) const {
+    if (weight < other.weight) return true;
```
Comment (csukuangfj) [inline k2/csrc/determinize.cc:40]: What is the return value if `weight == other.weight` ??

File: k2/csrc/determinize.cc [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+    if (inserted) {
+      *state_id = cur_output_state_++;
+      return true;
+    } else {
+      *state_id = p.first->second;
+      return false;
+    }
+  }
+
+  size_t size() const { return cur_output_state_; }
```
Comment (csukuangfj) [inline k2/csrc/determinize.cc:171]: why is it `size_t` ?

File: k2/csrc/fsa.h [hunk 8]
```diff
@@ -148,15 +144,43 @@ struct DenseFsa {
...
 };
 
-using StatePair = std::pair<StateId, StateId>;
+/*
+  This demonstrates an interface for a deterministic FSA or FST; it's similar
+  to Kaldi's DeterministicOnDemandFst class.  It can be used for things like
+  language models.  Actually we'll template on types like this.  There is no
+  need to actually inherit from this class.  */
+class DeterministicGenericFsa {
+
```
Comment (csukuangfj) [inline k2/csrc/fsa.h:161]: public:

is missing?

File: k2/csrc/fsa.h [hunk 9]
```diff
@@ -148,15 +144,43 @@ struct DenseFsa {
...
+  int32_t Start();
+
+
+  bool LookupArc(int32_t cur_state,
```
Comment (csukuangfj) [inline k2/csrc/fsa.h:165]: Why are there no `const` modifiers for `Get` ?

Comment (danpovey) [reply k2/csrc/fsa.h:165]: Because this is an interface for a possibly dynamic object.

File: k2/csrc/determinize.cc [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+
+ private:
+
+  int32_t cur_output_state_ { 0 };
+  std::unordered_map<std::pair<uint64_t, uint64_t>, int32_t, DetStateVectorHasher> map_;
```
Comment (qindazhu) [inline k2/csrc/determinize.cc:176]: `DetStateHasher`?

File: k2/csrc/determinize.cc (hunk 6 above)
Comment (danpovey) [reply k2/csrc/determinize.cc:40]: Sorry, I didn't realize I had included this code, it is not finished.  I have to get back to this.

File: k2/csrc/determinize.cc (hunk 10 above)
Comment (danpovey) [reply k2/csrc/determinize.cc:176]: Yeah, might do that.  This code needs to be finished.

File: k2/csrc/determinize.cc [hunk 11]
```diff
@@ -0,0 +1,230 @@
+// k2/csrc/fsa_algo.cc
```
Comment (csukuangfj) [inline k2/csrc/determinize.cc:1]: The filename should be changed.

File: k2/csrc/determinize.cc [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+    uint64_t a = d.base_state + 17489 * d.seq_len,
+        b = d.base_state * 103979  + d.seq_len;
+
+    // We choose an arbitrary DetStateElement (the first one in the list) to
+    // read the symbol sequence from; the symbol sequence will be the same no
+    // matter which element we choose to trace back.
+    DetStateElement *elem = d.head;
+    int32_t seq_len = d.seq_len;
+    for (int32_t i = 0; i < seq_len; i++) {
+      a = elem->symbol + 102299 * a;
```
Comment (csukuangfj) [inline k2/csrc/determinize.cc:195]: Is there an error inside the `for` loop?

`elem` is never updated.

----

It should be `++i`, not `i++`.

**Comment by danpovey** [comment]
I will fix these issues, but best to look again much later when that code
is finished.
Let's merge anyway though- I don't want to go to the trouble of removing
this part just yet.

**Comment by danpovey** [comment]
Merging to avoid future conflicts, although not everything is resolved here.

**Pull request merged**